umh-core/pkg/fsm/agent/actions.go

```diff
@@ -0,0 +1,170 @@
+	if a.ObservedState.AgentInfo.Status == agent_service.AgentStatusError {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// checkAndUpdateConfig checks for updated configuration and applies any changes
+// to the agent service. This ensures the agent is always running with the latest config.
+func (a *AgentInstance) checkAndUpdateConfig(ctx context.Context, tick uint64) error {
+	a.baseFSMInstance.GetLogger().Debugf("Checking for configuration updates...")
+
+	// In a real implementation, we would use the ConfigManager instead of loading directly from file
+	// The ConfigManager is typically passed to the FSM when it's created
+	if a.configManager == nil {
+		// If no config manager is available, we can't check for updates
+		a.baseFSMInstance.GetLogger().Warnf("No config manager available, skipping config update check")
+		return nil
+	}
+
+	// Get the latest config from the manager
+	fullConfig, err := a.configManager.GetConfig(ctx, tick)
+	if err != nil {
+		return fmt.Errorf("failed to get latest config: %w", err)
+	}
+
+	// Extract the agent config
+	latestConfig := fullConfig.Agent
+
+	// Check if the config has changed
+	if !reflect.DeepEqual(a.config, latestConfig) {
+		a.baseFSMInstance.GetLogger().Infof("Detected configuration change, updating agent...")
+
+		// Update our stored config
+		a.config = latestConfig
+
+		// Update location if changed
+		if !reflect.DeepEqual(a.ObservedState.AgentInfo.Location, latestConfig.Location) {
+			if err := a.service.UpdateLocation(ctx, latestConfig.Location); err != nil {
+				return fmt.Errorf("failed to update agent location: %w", err)
+			}
+		}
+
+		// Update release channel if changed
+		if a.ObservedState.AgentInfo.ReleaseChannel != latestConfig.ReleaseChannel {
+			if err := a.service.UpdateReleaseChannel(ctx, latestConfig.ReleaseChannel); err != nil {
+				return fmt.Errorf("failed to update agent release channel: %w", err)
+			}
+		}
+
+		// If more substantial changes that require reinitialization, we could set a flag
+		// to trigger a restart of the agent in the next reconcile cycle
+	}
+
+	return nil
+}
+
+// updateAgentObservedState updates the observed state of the agent by querying the service.
+func (a *AgentInstance) updateAgentObservedState(ctx context.Context, tick uint64) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(ComponentAgentInstance, "updateAgentObservedState", time.Since(start))
+	}()
+
+	// First, check for configuration updates
+	if err := a.checkAndUpdateConfig(ctx, tick); err != nil {
+		// We now propagate config errors upward to trigger degraded state
+		a.baseFSMInstance.GetLogger().Warnf("Failed to check for config updates: %v", err)
+		return fmt.Errorf("configuration update failed: %w", err)
+	}
+
+	// Get the current agent status
+	agentInfo, err := a.service.GetStatus(ctx)
+	if err != nil {
+		if err == agent_service.ErrAgentNotInitialized {
```

Comment:
the pattern here in s6 and benthos are slightly different:

if errors.Is(err, benthos_service.ErrServiceNotExist) {

or 

if s.baseFSMInstance.GetCurrentFSMState() == fsm.LifecycleStateCreating || s.baseFSMInstance.GetCurrentFSMState() == fsm.LifecycleStateToBeCreated {